Gleam-from-Elm cheatsheet (#109)
gleam-lang/website: cheatsheets/gleam-for-elm-users.md

- [Variables](#variables)
  - [Variables type annotations](#variables-type-annotations)
- [Constants](#constants)
- [Functions](#functions)
  - [Function type annotations](#function-type-annotations)
  - [Function heads](#function-heads)
  - [Function overloading](#function-overloading)
  - [Referencing functions](#referencing-function)
  - [Calling anonymous functions](#calling-anonymous-functions)
  - [Labelled arguments](#labelled-arguments)
- [Modules](#modules)
  - [Exports](#exports)
- [Blocks](#blocks)
- [Data types](#data-types)
  - [Numbers](#numbers)
  - [Strings](#strings)
  - [Tuples](#tuples)
  - [Records](#records) TODO
  - [Lists](#lists)
  - [Dicts](#dicts)
- [Operators](#operators)
- [Type aliases](#type-aliases)
- [Custom types](#custom-types) TODO
  - [Maybe](#maybe)
  - [Result](#result)
- [Commands](#commands) TODO
- [Architecture](#architecture) TODO
- [Talking to other languages](#talking-to-other-languages) TODO
- [Package management](#package-management)
- [Implementation](#implementation)
- [Other concepts](#other-concepts)
  - [Atoms](#atoms)

## Overview

Elm and Gleam have similar goals of providing a robust and sound type system with an approachable

> set of features. Gleam started off with a similar syntax to Elm but as moved to use one that is more
> familiar to many developers.

Reviewer: ```suggestion
set of features.
```

An interesting bit of history, but not something I think belongs here. I wouldn't want any newcomer Elm syntax fans to feel let down by this change.

Where Elm compiles to Javascript, Gleam initially aimed to compile to Erlang. Where Elm is best
suited for front-end browser based applications, Gleam initially targets a back-end ecosystem.

There is work in progess to compile Gleam to Javascript which opens up the possibility for front-end
development.

One area in which Elm and Gleam differ in their approach is where Elm does not provide easy access
to defining foreign-function-interfaces for interacting with Javascript libraries. In contrast to
this, Gleam makes it easy to define this inferfaces for using Erlang libraries.

## Comments

#### Elm

In Elm, single line comments are written with a `--` prefix and multiline comments are written with `{- -}` and `{-| -}` for documentation comments.

```elm
-- Hello, Joe!

{- Hello, Joe! 
   This is a multiline comment.
-}

{-| Determine the length of a list.
    length [1,2,3] == 3
-}
length : List a -> Int
length xs =
  foldl (\_ i -> i + 1) 0 xs
```

#### Gleam

In Gleam comments are written with a `//` prefix.

```rust
// Hello, Joe!
```

> Comments starting with `///` are used to document the following statement. Comments starting with `////` are used to document the current module.

Reviewer: ```suggestion
Comments starting with `///` are used to document the following function, constant, or type definition. Comments starting with `////` are used to document the current module.
```

There was some confusion around the word "statement" in discord the other day. I am hoping this language is clearer.

```rust
//// This module is very important.

/// The answer to life, the universe, and everything.
const answer: Int = 42
```

> There are no multiline commends in Gleam.

Reviewer: ```suggestion
There are no multiline comments in Gleam.
```

## Variables

#### Elm

In Elm, you assign variables in let-blocks and you cannot re-assign variables within a let-block. 

You also cannot create a variable with the same name as a variable from a higher scope.

```Elm
let
    size = 50
    size = size + 100 -- Compile Error!
in
```

#### Gleam

Gleam has the `let` keyword before its variable names. You can re-assign variables and you can

> shadow variables from other scopes.

Reviewer: ```suggestion
shadow variables from other scopes. This does not mutate the previously assigned value.
```

    42
```

#### Gleam

In Gleam constants can be created using the `const` keyword.

```rust
const the_answer = 42

pub fn main() {
  the_answer
}
```

Gleam constants can be referenced from other modules.

```rust
// in file other_module.gleam
pub const the_answer: Int = 42
```

```rust
import other_module

fn main() {
  other_module.the_answer
}
```

## Functions

#### Elm

In Elm, functions are defined as declarations that have arguments, or by assigning anonymous functions to variables.

```elm
sum x y =
  x + y

mul =
  \x y -> x * y

mul 3 2
-- 6
```

#### Gleam

Gleam's functions are declared using a syntax similar to Rust or JavaScript. Gleam's anonymous functions are declared using the `fn` keyword.

```rust
pub fn sum(x, y) {
  x + y
}

let mul = fn(x, y) { x * y }
mul(1, 2)
```

### Function type annotations

#### Elm

In Elm, functions can **optionally** have their argument and return types annotated. These type annotations will always be checked by the compiler and throw a compilation error if not valid. The compiler will still type check your program using type inference if annotations are omitted.

```elm
sum : number -> number -> number
sum x y = x + y

> @spec mul(number, number) :: boolean # no Elixir compile error

Reviewer: ```suggestion
```

mul : number -> number -> Bool -- Compile error
mul x y = x * y
```

#### Gleam

All the same things are true of Gleam though the type annotations go inline in the function
declaration, rather than above it.

```rust
pub fn add(x: Int, y: Int) -> Int {
  x + y
}

pub fn mul(x: Int, y: Int) -> Bool { // compile error, type mismatch
  x * y
}
```

> ### Function heads
> 
> Neither Elm nor Gleam support multiple function heads like Erlang, Elixir or Haskell. Both languages
> would expect a case-statement to be used to handle variations in data being supplied to the
> function.
> 
> ### Function overloading
> 
> Neither Elm nor Gleam support function overloading.
> 
> ### Referencing functions
> 
> Both Elm and Gleam have a single namespace for value and functions within a module, so there
> is no need for a special syntax to assign a module function to a variable.
> 
> ### Calling anonymous functions
> 
> Neither Elm nor Gleam require special syntax for calling anonymous functions.
> 

Reviewer: Maybe we can remove this bit? Not sure Elm programmers would be expecting these features. What do you think?

```suggestion
```

Author: Yeah, makes total sense. This a hang over from the original Elixir layout and it seemed odd to delete valid information but I can see it is potentially confusing if someone knows little about these things.

  y
}
```

## Data types

### Numbers

Both Elm and Gleam support Int and Float as separate number types. 

#### Gleam

Operators in Gleam as not generic over Int and Float so there are separate symbols for Int and Float operations. For example, `+` adds Ints together whilst `+.` adds Floats together. The pattern of the additional `.` extends to the other common operators.

Additionally, underscores can be added to both Ints and Floats for clarity.

```rust
const oneMillion = 1_000_000
const twoMillion = 2_000_000.0
```

### Strings

In both Elm and Gleam all strings support unicode. Gleam uses UTF-8 binaries. Elm compiles to Javascript which uses UTF-16 for its strings.

Both languages use double quotes for strings.

#### Elm

```elm
"Hellø, world!"
```

#### Gleam

```rust
"Hellø, world!"
```

### Tuples

Tuples are very useful in both Elm and Gleam as they're the only collection data type that allows mixed types in the collection.

#### Elm

In Elm, tuplies are limited to only 2 or 3 entries. It is recommended to use records for more larger numbers of entries.

```elm
myTuple = ("username", "password", 10)
(_, password, _) = myTuple
```

#### Gleam

> There is no limit to the number of entries in Gleam tuples.

Reviewer: ```suggestion
There is no limit to the number of entries in Gleam tuples, but records are still recommended as giving names to fields adds clarity.
```

Like Elm, there is no map literal syntax in Gleam, and you cannot pattern match on a map.

#### Elm

```Dict
import Dict

Dict.fromList [ ("key1", "value1"), ("key2", "value2") ]
Dict.fromList [ ("key1", "value1"), ("key2", 2) ] -- Compile error
```

#### Gleam

```rust
import gleam/map

map.from_list([#("key1", "value1"), #("key2", "value2")])
map.from_list([#("key1", "value1"), #("key2", 2)]) // Type error!
```

## Type Aliases

Elm uses type alises to define the layout of records. Gleam uses Custom Types to achieve a similar
result.

Gleam's custom types allow you to define a collection data type with a fixed number of named fields, and the values in those fields can be of differing types.

#### Elm

```elm
type alias Person =
 { name : String
 , age : Int 
 }

person = Person "Jake" 35
name = person.name
```

#### Gleam

Gleam's custom types can be used in much the same way. At runtime, they have a tuple representation and are compatible with Erlang records.

```rust
type Person {
  Person(name: String, age: Int)
}

let person = Person(name: "Jake", age: 35)
let name = person.name

> ```

Reviewer: Not sure what is happening here, GitHub's syntax highlighting seems to have broken?
